lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -1254,23 +1255,33 @@ def _iter_modules_impl(paths, prefix=''):
         prefix = to_native(prefix)
     # yield (module_loader, name, ispkg) for each module/pkg under path
     # TODO: implement ignore/silent catch for unreadable?
-    for b_path in map(to_bytes, paths):
-        if not os.path.isdir(b_path):
+    # Mostly based off the logic in the builtin pkgutil.iter_modules importer.
+    # Only difference is we don't check if dirs contains __init__.
+    # https://github.com/python/cpython/blob/fda056e64bdfcac3dd3d13eebda0a24994d83cb8/Lib/pkgutil.py#L130-L168
+    for path in paths:
+        if not os.path.isdir(path):
             continue
-        for b_basename in sorted(os.listdir(b_path)):
-            b_candidate_module_path = os.path.join(b_path, b_basename)
-            if os.path.isdir(b_candidate_module_path):
+
+        yielded = set()
+        for basename in sorted(os.listdir(path)):
+            modname = inspect.getmodulename(basename)
+            if modname == '__init__' or modname in yielded:
```

Comment:
How about we generalize this a little more?
```suggestion
        yielded = {'__init__'}
        for basename in sorted(os.listdir(path)):
            modname = inspect.getmodulename(basename)
            if modname in yielded:
```